chain/syncer_integration_test.go

```diff
@@ -0,0 +1,617 @@
+
+	fakeChildParams.Nonce = uint64(2)
+	forklink1blk3 := th.RequireMkFakeChild(t, fakeChildParams)
+	forklink1 := th.RequireNewTipSet(t, forklink1blk1, forklink1blk2, forklink1blk3)
+
+	fakeChildParams.Parent = forklink1
+	fakeChildParams.Nonce = uint64(0)
+	forklink2blk1 := th.RequireMkFakeChild(t, fakeChildParams)
+
+	fakeChildParams.Nonce = uint64(1)
+	forklink2blk2 := th.RequireMkFakeChild(t, fakeChildParams)
+
+	fakeChildParams.Nonce = uint64(2)
+	forklink2blk3 := th.RequireMkFakeChild(t, fakeChildParams)
+	forklink2 := th.RequireNewTipSet(t, forklink2blk1, forklink2blk2, forklink2blk3)
+
+	fakeChildParams.Nonce = uint64(0)
+	fakeChildParams.Parent = forklink2
+	forklink3blk1 := th.RequireMkFakeChild(t, fakeChildParams)
+
+	fakeChildParams.Nonce = uint64(1)
+	forklink3blk2 := th.RequireMkFakeChild(t, fakeChildParams)
+	forklink3 := th.RequireNewTipSet(t, forklink3blk1, forklink3blk2)
+
+	_ = requirePutBlocks(t, blockSource, forklink1.ToSlice()...)
+	_ = requirePutBlocks(t, blockSource, forklink2.ToSlice()...)
+	forkHead := requirePutBlocks(t, blockSource, forklink3.ToSlice()...)
+	err = syncer.HandleNewTipset(ctx, forkHead)
+	require.NoError(t, err)
+	requireHead(t, chainStore, forklink3)
+
+	// Put blocks in global IPLD blockstore
+	// TODO #2128 make this cleaner along with broad test cleanup.
+	bs := bstore.NewBlockstore(r.Datastore())
+	cst := &hamt.CborIpldStore{Blocks: bserv.New(bs, offline.Exchange(bs))}
+	requirePutBlocksToCborStore(t, cst, dstP.genTS.ToSlice()...)
+	requirePutBlocksToCborStore(t, cst, dstP.link1.ToSlice()...)
+	requirePutBlocksToCborStore(t, cst, dstP.link2.ToSlice()...)
+	requirePutBlocksToCborStore(t, cst, forklink1.ToSlice()...)
+	requirePutBlocksToCborStore(t, cst, forklink2.ToSlice()...)
+	requirePutBlocksToCborStore(t, cst, forklink3.ToSlice()...)
+
+	// Shut down store, reload and wire to syncer.
+	loadSyncer, blockSource := loadSyncerFromRepo(t, r, dstP)
+
+	// Test that the syncer can't sync a block on the old chain
+	// without getting old blocks from network. i.e. the repo is trimmed
+	// of non-heaviest chain blocks
+	cids3 := requirePutBlocks(t, blockSource, dstP.link3.ToSlice()...)
+	err = loadSyncer.HandleNewTipset(ctx, cids3)
+	assert.Error(t, err)
+
+	// Test that the syncer can sync a block on the heaviest chain
+	// without getting old blocks from the network.
+	fakeChildParams.Parent = forklink3
+	forklink4blk1 := th.RequireMkFakeChild(t, fakeChildParams)
+	forklink4 := th.RequireNewTipSet(t, forklink4blk1)
+	cidsFork4 := requirePutBlocks(t, blockSource, forklink4.ToSlice()...)
+	err = loadSyncer.HandleNewTipset(ctx, cidsFork4)
+	assert.NoError(t, err)
+}
+
+// Syncer handles MarketView weight comparisons.
+// Current issue: when creating miner mining with addr0, addr0's storage head isn't found in the blockstore
+// and I can't figure out why because we pass in the correct blockstore to createStorageMinerWithpower.
+func TestTipSetWeightDeep(t *testing.T) {
```

Comment:
I have made no attempt to port this sole remaining test. Likely it can be done so, possibly with a custom `StateBuilder` given to the chain.Builder which defines a different `Weight` function. Left for a follow-up. Or maybe that's unnecessary. Or maybe it should remain an integration test (but we can still rewrite it to use chain.Builder).

Reply:
As mentioned above we could also turn this into a FAST test, though that might be considerable work.